### DISCUSSION TITLE
Stitch scanned large music sheets A4 to A3

### DISCUSSION THREAD

**Discussion by VadeInPace** [ImageMagick/ImageMagick#7839] [Help]
### ImageMagick version

7.1.1-4 Q16-HDRI x86_64

### Operating system, version and so on

Linux Opensuse 15.5

### Description

Hello there,
Last year in March I got a tremendous help in this forum from snibgo and succeed to stitch 2 A4 pages in one A3 from few old music books. See [here](https://github.com/ImageMagick/ImageMagick/discussions/6176)
I need now to use the script again but I'd guess some parameters are not good for this scans and I got shifts between top and bottom lines in the A3 pages.
Unfortunately I am not skilled enough to fix it.
I can provide the script and some pages, please, let me know how.
I attach a bad result image.
Many thanks for your help
![com37](https://github.com/user-attachments/assets/0aee5279-2c60-49da-8433-8b1e80f70b3b)

**Comment by snibgo** [comment]
Please show the input images for that output, and the script you used.

**Reply by VadeInPace** [reply]
Top A4
![p37](https://github.com/user-attachments/assets/ef5b5931-904c-4f47-b4c8-7a5f1faa2266)

Bottom A4

![ob37](https://github.com/user-attachments/assets/f94fc01f-6e5d-47fc-837e-e8473d096673)

**Reply by VadeInPace** [reply]
Script:

```
#!/bin/bash

# Top image dir - Add the dir with / - This is the top of the page
TopImgDir="oddtop/"
# Bottom image dir - Add the dir with / - This is the bottom of the page
BotImgDir="oddbot/"
# Composed image dir - Add the dir with / - This is the final page
ComImgDir="oddcomp2"

# Find bottom image prefix
onefile="$(ls "$BotImgDir" | grep '.png' | head -1)"
BotImgPre="${onefile%%[0-9]*}"
number=1

# Loop in the Top image directory 
for file in "$TopImgDir"*.png
do
# Debug: Stop after n images
#if [ "$number" -gt 20 ]; then exit; fi
	ImgTop="$file"
# Extract the file number
	number="${ImgTop//[^0-9]}"
# Corresponding bottom file
	ImgBot="$BotImgDir""$BotImgPre""$number"".png"

# Composed image
	ImgCom="$ComImgDir""/com""$number"".png"
echo "$ImgTop"
echo "$ImgBot"
echo "$ImgCom" 

# Find TrimY, the amount we will trim from the top of music2.

str="$( ${BIMG7}magick \
  $ImgBot \
  -morphology Dilate Rectangle:5x1 \
  -morphology Erode Rectangle:100x1 \
  -scale "1x!" \
  -auto-level \
  -threshold 98% \
  -define "identify:locate=maximum" \
  -define "identify:limit=1" \
  info: )"

echo str=$str

TrimY=$( echo $str |cut --only-delimited --delimiter=, --fields=2 )

echo TrimY=$TrimY

# Get c1.png, the bottom-left corner of Imgtop, recording the Y-offset of this in TopY.
# From Imgbot, crop the first bar of one stave, which we will search for in c1.png.
# For extra performance we trim white space from the crop,
# after blurring and thresholding to de-noise.
# Store that in c2t.png.
#
# 670x280 captures the first bar of one stave.
# bellow: +gravity -crop widthxhightx+0+$TrimY
# +gravity -crop 800x700+0+$TrimY \ good for ~1670x1640 images
#

declare $(${BIMG7}magick \
  $ImgTop \
  -gravity SouthWest -crop 50x50%+0+0 \
  -format "TopY=%Y\n" +write info: \
  +repage \
  -write c1.png \
  +delete \
  $ImgBot \
  +gravity -crop 900x700+0+$TrimY \
  \( +clone \
     -blur 0x3 -threshold 50% -trim \
     -format "CropX=%X\nCropY=%Y\n" +write info: \
     -set option:CropIt %wx%h%O \
     +delete \
  \) \
  -crop %[CropIt] +repage \
  +write c2t.png \
  -format "CropIt=%[CropIt]\n" \
  info:)

echo TopY=$TopY CropX=$CropX CropY=$CropY CropIt=$CropIt

# Search for position of c2t in c1 using process module 'srchimg', takes about 1s.

str=$( ${BIM7DEV}magick \
  c1.png \
  c2t.png \
  -process 'srchimg file stdout' \
  NULL: )

echo str=$str

str=$( echo $str |cut --only-delimited --delimiter=" "  --fields=3 )
echo str=$str

SrchX=$( echo $str |cut --only-delimited --delimiter=,  --fields=1 )
SrchY=$( echo $str |cut --only-delimited --delimiter=,  --fields=2 )

echo fast SrchX=$SrchX SrchY=$SrchY


# Trim ImgBot and merge it over ImgTop.
# Make it horizontal and adjust the size
# Bellow: -background White -gravity Center -extent 2610x3710 \
# Porret 
#	-background White -gravity Center -extent 1700x2410 \
#	-gravity Center -crop 1700x2410+0+0 +repage \

	${BIMG7}magick \
	$ImgTop \
	\(  $ImgBot -crop x+0+$TrimY +repage \
	-repage "+%[fx:$SrchX-($CropX)]+%[fx:$TopY+$SrchY+$TrimY-($CropY)]" \
	\) \
	-compose Over -layers merge \
	+repage \
	-background White -gravity Center -extent 1870x2510 \
	-gravity Center -crop 1870x2520+0+0 +repage \
	$ImgCom

done
```

**Comment by snibgo** [comment]
I suppose the problem you want to solve is on the right side of the output image, the five ends of stave lines. These come from the top image, and are because your final magick command merges the bottom image over the top image with a y-offset that is reasonable, but also with an x-offset of "$SrchX-($CropX)". So the bottom image is shifted to the left in the output, which reveals a bit of the right side of the top image. Replacing that offset with "0" fixes the problem, and also aligns the staves from both images more closely.

So change the line...
```
	-repage "+%[fx:$SrchX-($CropX)]+%[fx:$TopY+$SrchY+$TrimY-($CropY)]" \
```
... to ...
```
	-repage "+0+%[fx:$TopY+$SrchY+$TrimY-($CropY)]" \
```

**Reply by VadeInPace** [reply]
Thank you snibgo, I get a better result.
However, some pages are not good where there is text and where there is vertical line.
For example
Top
![p21](https://github.com/user-attachments/assets/5970664b-3700-4636-9354-fbd979962fb3)
Bottom
![ob21](https://github.com/user-attachments/assets/b547d559-31dc-425f-9ead-2ea082be9109)

**Reply by VadeInPace** [reply]
Another bad result
$107bis should have 4 staves.
Top
![p33](https://github.com/user-attachments/assets/9ffdd784-ea90-4563-93dd-e2501c72d69f)
Bottom
![ob33](https://github.com/user-attachments/assets/0d105b2d-c415-41fe-9d28-ed5b2e5e672c)

**Reply by VadeInPace** [reply]
There are more like that after.
Thanks for your help

**Reply by VadeInPace** [reply]
Any help about the above?
Thanks

**Comment by VadeInPace** [comment]
Hi snibgo,
Could you help me about the sheets above ?
If you need more info let me know.
Many thanks

**Comment by snibgo** [comment]
The following bash script gives accurate alignment for each image pair show on this forum thread. From the second image, we trim the top, left and right sides, and save the image in a file `musicTmp.miff`. Then we crop the top 30% of the image, and search for that in the first image. This gives the coordinates of the trimmed second image in the first image.

```
FileA=music107A.png
FileB=music107B.png

declare ImLine=$( magick \
 ${FileB} \
 -fuzz 30% \
 -define trim:edges=north,east,west \
 -trim +repage \
 +write musicTmp.miff \
 -crop %wx%[fx:h*0.3]+0+0 \
  ${FileA} \
  +swap \
  -process 'srchimg nocrop canvasOffset file stdout' \
  NULL: )

echo ImLine=${ImLine}

FndX=$( echo ${ImLine} | cut --delimiter ' ' --fields 3 | cut --delimiter ',' --fields 1 )
FndY=$( echo ${ImLine} | cut --delimiter ' ' --fields 3 | cut --delimiter ',' --fields 2 )

echo FndX=${FndX} FndY=${FndY}

magick \
  musicTmp.miff \
  -repage +${FndX}+${FndY} \
  ${FileA} \
  +swap \
  -compose Over -layers Mosaic \
  output.png
```
Like my previous code, this uses my [Process module: multi-scale image search](http://im.snibgo.com/customim.htm#srchimg).

**Comment by VadeInPace** [comment]
Thanks a lot snibgo, I appreciate
This is working very well, even better in general that your previous method/code.

However, due to my scan, almost all top edges of the bottom images have a grey part of around 1/10 of the high of the image. You can see in the first bottom image above. In fact, in the process, the bottom image is in full size and the top page is cropped.
Is is possible to trim around 1/5 to 1/4 of the to part of the image and then search the top part ?
If you tell me the place in the code I'll adjust it. 
I'd guess this is in the command -crop %wx%[fx:h*0.3]+0+0 \
Thanks

**Comment by snibgo** [comment]
Just altering the `-crop` won't work.

You could remove the final `-swap`, so instead of getting the top of FileB you get the bottom of FileA, which is less bad. Or you could make the top of musicTmp.miff in the final command transparent. Or you could do an alpha blend in the overlap, so the opacity of musicTmp varies from fully transparent at the top to full opaque at the bottom of FileA. This relies on the alignment being excellent, so there is no ghosting at partial transparency.
```
declare OverHt=$( magick.exe \
  -ping \
  ${FileA} \
  -format %[fx:h-${FndY}] \
  info: )

echo OverHt=${OverHt}

magick.exe \
  musicTmp.miff \
  \( +clone \
     -sparse-color bilinear "0,0,Black %[fx:w-1],0,Black 0,${OverHt},White" \
  \) \
  -alpha off \
  -compose CopyOpacity -composite \
  -repage +${FndX}+${FndY} \
  ${FileA} \
  +swap \
  -compose Over -layers Mosaic \
  output2.png
```

**Comment by VadeInPace** [comment]
Thank you snibgo,
However, I don't know where to insert this code, it uses only the top image.
Let me know.

**Comment by snibgo** [comment]
You can add this to the end of my previous script.

It uses the variable FileA and the file musicTmp.miff, both of which are setup in the script.

**Reply by VadeInPace** [reply]
Almost done!
Better than previously but 3 of them have issue, not exactly overlaid.
For example
TOP
![evetop34](https://github.com/user-attachments/assets/d6feef91-2455-47fb-b786-b0f16dcac598)

BOTTOM
![evebot34](https://github.com/user-attachments/assets/ffb8b2bf-9db7-4226-8db2-081f8c53b739)

**Comment by snibgo** [comment]
If you view musicTmp.miff, you will see it has not been fully trimmed. Increase the fuzz to 50% to fix this.

**Reply by VadeInPace** [reply]
Very good now!
May I be a bit perfectionist?
The top images have a grey part on the right edge on white background. Is is possible to easily remove it?
If not easy, I'll leave it.
For example
TOP
![evetop12](https://github.com/user-attachments/assets/cfe59a1b-e9ca-4fda-b13f-7f9d08acd21a)
BOTTOM
![evebot12](https://github.com/user-attachments/assets/5c79e386-df3a-42fd-985f-2f518eebde06)
Many thanks

**Comment by snibgo** [comment]
To fix uneven lighting of a scan of dark ink on light paper, we can first create an image that represents the paper with no ink. Then we divide the input image by this inkless paper image, so the result has white paper.

We make the inkless paper by making the ink transparent, then filling the transparent holes (aka "infill").

To make the ink transparent, we make pixels darker than 50% transparent, and enlarging those areas with morphology.

```
magick.exe \
  music19A.png -write mpr:INP \
  -fuzz 50% -transparent Black \
  \( +clone \
     -alpha extract \
     -morphology erode disk:2.0 \
  \) \
  -alpha off \
  -compose CopyOpacity -composite \
  -process 'fillholes lsr 10 window_radius 2 copy window' \
  mpr:INP \
  -compose DivideDst -composite \
  out.png
```
This takes 18 seconds on my laptop. Almost all that time is filling the holes. Tweaking the "fillholes" parameters can accelerate this, but might reduce quality. See [Filling holes](http://im.snibgo.com/fillholes.htm).

If my process module "fillholes" is not available, the job can be done with "seamless_blend":
```
magick.exe \
  music19A.png -write mpr:INP \
  -fuzz 50% -transparent Black \
  \( +clone \
     -alpha extract \
     -morphology erode disk:2.0 \
  \) \
  -alpha off \
  -compose CopyOpacity -composite \
  \( +clone -alpha extract -negate +write mpr:ALP +delete \) \
  \( -clone 0 -alpha off -fill Black -colorize 100 \) \
  mpr:ALP \
  -verbose \
  -define compose:args=10000x1e-7+1000 \
  -compose seamless_blend -composite \
  mpr:INP \
  -compose DivideDst -composite \
  out2.png
```
This takes 3 minutes on my laptop. Tweaking the "compose:args" can accelerate this, but might reduce quality.

**Reply by VadeInPace** [reply]
Thanks a lot, now the sheets look like a A3 scanner. Very good!
FYI:
There are 2 way to insert the last code to remove grey area, or at the beginning for top then bottom or at the end when the page is stitched. 
The first, at beginning, take 7.6 seconds/page calculated with 37 pages batch.
The last, at the end, take 1 mn/page.

For reference and sharing, here the full code to batch stitch top A4 scan and bottom A4 scan from A3 original.
Please could you look at it and improve it if necessary. 
Many thanks for your valuable help.
```
#!/bin/bash

# Top image dir - Add the dir with / - This is the top of the page
TopImgDir="/path/to/oddtop/"
#TopImgDir="/path/to/evetop/"
# Bottom image dir - Add the dir with / - This is the bottom of the page
BotImgDir="/path/to/oddbot/"
#BotImgDir="/path/to/evebot/"
# Composed image dir - Add the dir with / - This is the composed page
ComImgDir="/path/to/oddcomp/"
#ComImgDir="/path/to/evecomp/"

# Find bottom image prefix
onefile="$(ls "$BotImgDir" | grep '.png' | head -1)"
BotImgPre="${onefile%%[0-9]*}"

# Loop in the Top image directory 
for file in "$TopImgDir"*.png
do
	ImgTop="$file"
# Extract the file number
	number="${ImgTop//[^0-9]}"
# Corresponding bottom file
	ImgBot="$BotImgDir""$BotImgPre""$number"".png"

# Composed image
	ImgCom="$ComImgDir""com""$number"".png"

#echo "$ImgTop"
#echo "$ImgBot"
#echo "$ImgCom"

# Remove grey area
magick \
  ${ImgBot} -write mpr:INP \
  -fuzz 50% -transparent Black \
  \( +clone \
     -alpha extract \
     -morphology erode disk:2.0 \
  \) \
  -alpha off \
  -compose CopyOpacity -composite \
  -process 'fillholes lsr 10 window_radius 2 copy window' \
  mpr:INP \
  -compose DivideDst -composite \
  IntBot.png

magick \
  ${ImgTop} -write mpr:INP \
  -fuzz 50% -transparent Black \
  \( +clone \
     -alpha extract \
     -morphology erode disk:2.0 \
  \) \
  -alpha off \
  -compose CopyOpacity -composite \
  -process 'fillholes lsr 10 window_radius 2 copy window' \
  mpr:INP \
  -compose DivideDst -composite \
  IntTop.png

# Start stitching process
declare ImLine=$( magick \
 IntBot.png \
 -fuzz 50% \
 -define trim:edges=north,east,west \
 -trim +repage \
 +write musicTmp.miff \
 -crop %wx%[fx:h*0.3]+0+0 \
  IntTop.png \
  +swap \
  -process 'srchimg nocrop canvasOffset file stdout' \
  NULL: )

echo ImLine=${ImLine}

FndX=$( echo ${ImLine} | cut --delimiter ' ' --fields 3 | cut --delimiter ',' --fields 1 )
FndY=$( echo ${ImLine} | cut --delimiter ' ' --fields 3 | cut --delimiter ',' --fields 2 )

echo FndX=${FndX} FndY=${FndY}

magick \
  musicTmp.miff \
  -repage +${FndX}+${FndY} \
  IntTop.png \
  +swap \
  -compose Over -layers Mosaic \
  Intermed_bin
  
  
declare OverHt=$( magick \
  -ping \
  IntTop.png \
  -format %[fx:h-${FndY}] \
  info: )

echo OverHt=${OverHt}

magick \
  musicTmp.miff \
  \( +clone \
     -sparse-color bilinear "0,0,Black %[fx:w-1],0,Black 0,${OverHt},White" \
  \) \
  -alpha off \
  -compose CopyOpacity -composite \
  -repage +${FndX}+${FndY} \
  IntTop.png \
  +swap \
  -compose Over -layers Mosaic \
  -background White -gravity Center -extent 1700x2400 \
  -gravity Center -crop 1700x2400+0+0 +repage \
  $ImgCom

done
```
